src/Juvix/Compiler/Tree/EvaluatorSem.hs

```diff
@@ -0,0 +1,389 @@
+        withExtendedTemp v m = do
+          ctx <- ask
+          withCtx (over evalCtxTemp (BL.cons v) ctx) m
+
+        goCaseBranch :: Value -> Bool -> Node -> Sem r' Value
+        goCaseBranch v bSave body
+          | bSave = withExtendedTemp v (eval' body)
+          | otherwise = eval' body
+
+        goSave :: NodeSave -> Sem r' Value
+        goSave NodeSave {..} = do
+          v <- eval' _nodeSaveArg
+          withExtendedTemp v (eval' _nodeSaveBody)
+
+printValue :: InfoTable -> Value -> Text
+printValue tab = \case
+  ValString s -> s
+  v -> ppPrint tab v
+
+valueToNode :: Value -> Node
+valueToNode = \case
+  ValInteger i -> mkConst $ ConstInt i
+  ValBool b -> mkConst $ ConstBool b
+  ValString s -> mkConst $ ConstString s
+  ValUnit -> mkConst ConstUnit
+  ValVoid -> mkConst ConstVoid
+  ValConstr Constr {..} ->
+    AllocConstr
+      NodeAllocConstr
+        { _nodeAllocConstrInfo = mempty,
+          _nodeAllocConstrTag = _constrTag,
+          _nodeAllocConstrArgs = map valueToNode _constrArgs
+        }
+  ValClosure Closure {..} ->
+    AllocClosure
+      NodeAllocClosure
+        { _nodeAllocClosureInfo = mempty,
+          _nodeAllocClosureFunSymbol = _closureSymbol,
+          _nodeAllocClosureArgs = map valueToNode _closureArgs
+        }
+
+hEvalIOEither ::
+  forall m.
+  (MonadIO m) =>
+  Handle ->
+  Handle ->
+  InfoTable ->
+  FunctionInfo ->
+  m (Either TreeError Value)
+hEvalIOEither hin hout infoTable funInfo = do
+  let x = do
+        v <- eval infoTable (funInfo ^. functionCode)
+        hRunIO hin hout infoTable v
+  let handleTrace = liftIO . hPutStrLn hout . printValue infoTable
+  liftIO
+    . runM
+    . runError @TreeError
+    . mapError toTreeError
+    . runOutputSem handleTrace
+    $ x
+
+-- | Interpret IO actions.
+hRunIO :: forall r. (Members '[Embed IO, Error EvalError, Output Value] r) => Handle -> Handle -> InfoTable -> Value -> Sem r Value
+hRunIO hin hout infoTable = \case
+  ValConstr (Constr (BuiltinTag TagReturn) [x]) -> return x
+  ValConstr (Constr (BuiltinTag TagBind) [x, f]) -> do
+    x' <- hRunIO hin hout infoTable x
+    let code =
+          CallClosures
+            NodeCallClosures
+              { _nodeCallClosuresInfo = mempty,
+                _nodeCallClosuresFun = valueToNode f,
+                _nodeCallClosuresArgs = valueToNode x' :| []
+              }
+    res <- eval infoTable code
+    hRunIO hin hout infoTable res
+  ValConstr (Constr (BuiltinTag TagWrite) [ValString s]) -> do
+    liftIO $ hPutStr hout s
+    return ValVoid
+  ValConstr (Constr (BuiltinTag TagWrite) [arg]) -> do
+    liftIO $ hPutStr hout (ppPrint infoTable arg)
+    return ValVoid
+  ValConstr (Constr (BuiltinTag TagReadLn) []) -> do
+    liftIO $ hFlush hout
+    s <- liftIO $ hGetLine hin
+    return (ValString s)
+  val ->
+    return val
+
+-- | Catch EvalError and convert it to TreeError.
+catchEvalErrorIO :: IO a -> IO (Either TreeError a)
```

Comment:
Duplicated and not needed for the effects-based version

Reply:
done